src/raftstore/store/store.rs

```diff
@@ -986,32 +962,28 @@ impl<T: Transport, C: PdClient> Store<T, C> {
     fn handle_gc_peer_msg(&mut self, msg: &RaftMessage) {
         let region_id = msg.get_region_id();
 
-        let mut need_remove = false;
-        let mut async_remove = true;
+        let mut job = None;
         if let Some(peer) = self.region_peers.get_mut(&region_id) {
+            assert_eq!(peer.peer, *msg.get_to_peer());
             // TODO: need checking peer id changed?
             let from_epoch = msg.get_region_epoch();
             if util::is_epoch_stale(peer.get_store().region.get_region_epoch(), from_epoch) {
                 // TODO: ask pd to guarantee we are stale now.
                 info!(
-                    "[region {}] peer {:?} receives gc message, remove",
+                    "[region {}] peer {:?} receives gc message, trying to remove",
                     region_id,
                     msg.get_to_peer()
                 );
-                need_remove = true;
-                peer.pending_remove = true;
-                async_remove = peer.get_store().is_initialized();
+                job = peer.maybe_destroy();
+                if job.is_none() {
+                    self.raft_metrics.message_dropped.applying_snap += 1;
+                    return;
+                }
```

Comment:
Why not call `job.execute(self)` at here?

Reply:
Same as above.